controllers/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -997,36 +996,45 @@ func TestDeleteNSTemplateSet(t *testing.T) {
 		require.Empty(t, result)
 		require.NoError(t, err)
 
-		// get the first namespace and check that it is not deleted
+		// get the first namespace and check that it has deletion timestamp
 		firstNSName := fmt.Sprintf("%s-dev", username)
-		AssertThatNamespace(t, firstNSName, r.Client).DoesExist()
+		AssertThatNamespace(t, firstNSName, r.Client).HasDeletionTimestamp()
 		// get the NSTemplateSet resource again, check it is not deleted and its status
 		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).
 			HasFinalizer().
 			HasConditions(UnableToTerminate("user namespace johnsmith-dev deletion was triggered but is not complete yet, something could be blocking ns deletion"))
 
-		// remove deletion timestamp, actual delete is now triggered
-		userNS := &corev1.Namespace{}
-		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: firstNSName}, userNS)
-		require.NoError(t, err)
-		userNS.DeletionTimestamp = nil
-		err = r.Client.Update(context.TODO(), userNS)
-		require.NoError(t, err)
+		// set MockDelete to nil
 		fakeClient.MockDelete = nil //now removing the mockDelete
 
-		//reconcile, deletion is triggered
+		//reconcile to check there is no change, ns still exists
 		result, err = r.Reconcile(context.TODO(), req)
 		require.Empty(t, result)
 		require.NoError(t, err)
-		// second reconcile triggered when ns is deleted
+
+		AssertThatNamespace(t, firstNSName, r.Client).HasDeletionTimestamp()
+		// get the NSTemplateSet resource again, check it is not deleted and its status
+		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).
+			HasFinalizer().
+			HasConditions(UnableToTerminate("user namespace johnsmith-dev deletion was triggered but is not complete yet, something could be blocking ns deletion"))
+
+		// actually delete ns
+		ns := &corev1.Namespace{}
+		err = r.Client.Get(context.TODO(),types.NamespacedName{Name: firstNSName},ns)
+		require.NoError(t, err)
+		err = r.Client.Delete(context.TODO(),ns)
+		require.NoError(t, err)
+
+		// deletion of ns would trigger another reconcile
 		result, err = r.Reconcile(context.TODO(), req)
 		require.Empty(t, result)
 		require.NoError(t, err)
 
 		// get the first namespace and check it IS deleted
 		AssertThatNamespace(t, firstNSName, r.Client).DoesNotExist()
 		// Check that nsTemplateSet is set to delete
-		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).DoesNotHaveFinalizer().HasConditions(Terminating())
+		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).
+			DoesNotHaveFinalizer().HasConditions(Terminating())
```

Comment:
Looks good!
So now we just need to delete the second namespace and check that the finalizer is finally removed from the NSTemplateSet.

Reply:
do we need to have two namespaces? Is it ok if the testcase only has one namespace associated with the user?